configs/module.go

```diff
@@ -491,3 +491,16 @@ func (m *Module) ProviderForLocalConfig(pc addrs.LocalProviderConfig) addrs.Prov
 	}
 	return addrs.NewLegacyProvider(pc.LocalName)
 }
+
+func (m *Module) ProviderForResource(r *Resource) addrs.LocalProviderConfig {
```

Comment:
👋 this is the function that started this PR, and specifically what I am looking for review on. Everything else is subject to change (feel free to comment on other changes - just understand that this is the key function that everything else sprang from).